v2/carbs/carbs.go

```diff
@@ -0,0 +1,300 @@
+package carbs
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"encoding/binary"
+	"fmt"
+	"io"
+
+	blocks "github.com/ipfs/go-block-format"
+	"github.com/ipfs/go-cid"
+	bs "github.com/ipfs/go-ipfs-blockstore"
+	"github.com/multiformats/go-multihash"
+
+	pb "github.com/cheggaaa/pb/v3"
+	car "github.com/ipld/go-car"
+	"github.com/ipld/go-car/util"
+	"golang.org/x/exp/mmap"
+)
+
+var errNotFound = bs.ErrNotFound
+
+// Carbs provides a read-only Car Block Store.
+type Carbs struct {
+	backing io.ReaderAt
+	idx     Index
+}
+
+var _ bs.Blockstore = (*Carbs)(nil)
+
+func (c *Carbs) Read(idx int64) (cid.Cid, []byte, error) {
+	bcid, data, err := util.ReadNode(bufio.NewReader(&unatreader{c.backing, idx}))
+	return bcid, data, err
+}
+
+// DeleteBlock doesn't delete a block on RO blockstore
+func (c *Carbs) DeleteBlock(_ cid.Cid) error {
+	return fmt.Errorf("read only")
+}
+
+// Has indicates if the store has a cid
+func (c *Carbs) Has(key cid.Cid) (bool, error) {
+	offset, err := c.idx.Get(key)
+	if err != nil {
+		return false, err
+	}
+	uar := unatreader{c.backing, int64(offset)}
+	_, err = binary.ReadUvarint(&uar)
+	if err != nil {
+		return false, err
+	}
+	cid, _, err := readCid(c.backing, uar.at)
+	if err != nil {
+		return false, err
+	}
+	return cid.Equals(key), nil
+}
+
+var cidv0Pref = []byte{0x12, 0x20}
+
+func readCid(store io.ReaderAt, at int64) (cid.Cid, int, error) {
+	var tag [2]byte
+	if _, err := store.ReadAt(tag[:], at); err != nil {
+		return cid.Undef, 0, err
+	}
+	if bytes.Equal(tag[:], cidv0Pref) {
+		cid0 := make([]byte, 34)
+		if _, err := store.ReadAt(cid0, at); err != nil {
+			return cid.Undef, 0, err
+		}
+		c, err := cid.Cast(cid0)
+		return c, 34, err
+	}
+
+	// assume cidv1
+	br := &unatreader{store, at}
+	vers, err := binary.ReadUvarint(br)
+	if err != nil {
+		return cid.Cid{}, 0, err
+	}
+
+	// TODO: the go-cid package allows version 0 here as well
+	if vers != 1 {
+		return cid.Cid{}, 0, fmt.Errorf("invalid cid version number: %d", vers)
+	}
+
+	codec, err := binary.ReadUvarint(br)
+	if err != nil {
+		return cid.Cid{}, 0, err
+	}
+
+	mhr := multihash.NewReader(br)
+	h, err := mhr.ReadMultihash()
+	if err != nil {
+		return cid.Cid{}, 0, err
+	}
+
+	return cid.NewCidV1(codec, h), int(br.at - at), nil
+}
+
+// Get gets a block from the store
+func (c *Carbs) Get(key cid.Cid) (blocks.Block, error) {
+	offset, err := c.idx.Get(key)
+	if err != nil {
+		return nil, err
+	}
+	entry, bytes, err := c.Read(int64(offset))
+	if err != nil {
+		fmt.Printf("failed get %d:%v\n", offset, err)
```

Comment:
I think this deserves a TODO; a library should never print directly to stdout/stderr